### PR TITLE
Add extensionAccessControl rule

### PR DESCRIPTION
**Pull request by calda**
The PR adds a new disabled-by-default `extensionAccessControl` rule, which lets users specify whether they want an extension's access control keyword to be placed on the extension itself, or on each individual body declaration.

`--extensionacl on-extension` (default)

```diff
- extension Foo {
-     public func bar() {}
-     public func baaz() {}
  }

+ public extension Foo {
+     func bar() {}
+     func baaz() {}
  }
```

`--extensionacl on-declarations`

```diff
- public extension Foo {
-     func bar() {}
-     func baaz() {}
-     internal func quux() {}
  }

+ extension Foo {
+     public func bar() {}
+     public func baaz() {}
+     func quux() {}
  }
```

### CODE REVIEW

File: Sources/FormattingHelpers.swift [hunk 1]
```diff
@@ -403,42 +403,162 @@ extension Formatter {
         }
         removeToken(at: index)
     }
+}
+
+/// Helpers for recursively traversing the declaration hierarchy
+extension Formatter {
+    /// Applies `mapDeclaration` to every recursive declaration from this formatter's tokens
+    func mapRecursiveDeclarations(mapDeclaration: (Declaration) -> Declaration) {
```
Comment (calda) [inline Sources/FormattingHelpers.swift:411]: These are some useful helpers that I pulled out of `organizeDeclarations`

File: Sources/FormattingHelpers.swift [hunk 2]
```diff
@@ -460,6 +595,15 @@ private extension Formatter {
         }
     }
 
+    /// The visibility of a declaration
+    enum Visibility: String {
```
Comment (calda) [inline Sources/FormattingHelpers.swift:599]: I pulled out the `visibility(of: Declaration)` logic from `organizeDeclarations` so I could use it in `extensionDeclarationVisibility`

**Comment by calda** [comment]
It looks like this builds with Swift 5.3 but not Swift 5.0. CI is building with Xcode 10, so it's using Swift 5.0. Is this something you're intentionally trying to support?

**Comment by nicklockwood** [comment]
In my own code I prefer the opposite approach (always hoisting the access level to the extension rather than specifying it per-element) since it reduces duplication and allows you to structure code like:

```swift
public struct Foo {
  public var bar: Int
}

// MARK: public methods
public extension {
  func x() {}
  func y() {}
}

// MARK: private methods
private extension {
  func z() {}
}
```

That doesn't mean I'm opposed to having a rule that does the opposite, but following the pattern of other such rules I'd probably be inclined to call it something like `hoistExtensionACL` and have a configuration option to toggle between hoisting and inlining access, similar to how `hoistPatternLet` works now.

Of course I don't expect you to do that work since it doesn't support your use-case. I just need to think whether it will be difficult to implement such a rule in future if this is merged now. The deprecation process for rules isn't quite as well defined as it is for options.

**Comment by nicklockwood** [comment]
> It looks like this builds with Swift 5.3 but not Swift 5.0. CI is building with Xcode 10, so it's using Swift 5.0. Is this something you're intentionally trying to support?

Yes. SwiftFormat needs to be able to to be built on every OS version that it supports (this is a Homebrew requirement). It currently supports back to macOS 13, which means it needs to use Swift 4.2 features only, and be able to compile with Xcode 10 or earlier (IIRC).

**Comment by calda** [comment]
> That doesn't mean I'm opposed to having a rule that does the opposite, but following the pattern of other such rules I'd probably be inclined to call it something like hoistExtensionACL and have a configuration option to toggle between hoisting and inlining access, similar to how hoistPatternLet works now.

I'm totally open to making this change. _"If all extension members have the same visibility, move that visibility onto the extension itself"_ seems like a totally reasonable inverse-option for this rule (and probably a better default for most users).

```diff
- extension Foo {
-   public func x() {}
-   public func y() {}
}

+ public extension Foo {
+   func x() {}
+   func y() {}
}
```

**Comment by calda** [comment]
Ah I see this in the README now:

> The SwiftFormat framework and command-line tool can be compiled using Swift 4.2 and above

Thanks for clarifying! I'll fix-up support for Swift 4.2

**Comment by calda** [comment]
@nicklockwood I updated this rule to be configurable with an `--extensionsacl` option:

`--extensionacl on-extension` (default)

```diff
- extension Foo {
-     public func bar() {}
-     public func baaz() {}
  }

+ public extension Foo {
+     func bar() {}
+     func baaz() {}
  }
```

`--extensionacl on-declarations`

```diff
- public extension Foo {
-     func bar() {}
-     internal var baaz: Int { 10 }
-     private func quux() {}
  }

+ extension Foo {
+     public func bar() {}
+     var baaz: Int { 10 }
+     private func quux()
  }
```

**Comment by nicklockwood** [comment]
@calda Amazing! Thank you

**Comment by nicklockwood** [comment]
@calda FYI: https://github.com/nicklockwood/SwiftFormat/commit/9d61be97f3afc9cb3ba88fb5549fa1affc889554

**Comment by nicklockwood** [comment]
@calda FYI: https://github.com/nicklockwood/SwiftFormat/commit/a2242da7ba8de721c70cf35cda666a88d5783654

**Comment by calda** [comment]
Thanks for these fixes! Extension access control is more complicated than I realized 😳

**Pull request merged**